utils/wasm-gen/src/config/syscalls/pointer_write.rs

```diff
@@ -0,0 +1,136 @@
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Entities describing config of pointer writes that `wasm-gen`
+//! includes into generated wasms. They're used to set values of some
+//! syscall parameters that are passed as pointer to the data.
+
+use arbitrary::{Result, Unstructured};
+use gear_wasm_instrument::syscalls::PtrType;
+use std::{collections::HashMap, mem::size_of, ops::RangeInclusive};
+
+pub use gear_wasm_instrument::syscalls::ParamType;
+
+/// Pointer writes config.
+///
+/// Determines what data should be written into the pointers of
+/// particular type. For example, it can be configured to write
+/// some data (generated in [`PointerWriteData::generate_data_to_write`])
+/// into all the pointers of type `*const/*mut Value` among syscall
+/// parameters.
+///
+/// # Note
+///
+/// This config will not work for [`PtrType::BufferStart`].
+#[derive(Debug, Clone)]
+pub struct PointerWritesConfig(HashMap<PtrType, Vec<PointerWrite>>);
+
+impl Default for PointerWritesConfig {
+    fn default() -> PointerWritesConfig {
+        let value_write_data = PointerWriteData::U128(0..=100_000_000_000);
+
+        const HASH_LEN: usize = size_of::<gsys::Hash>() / size_of::<i32>();
+
+        PointerWritesConfig(
+            [
+                (
+                    PtrType::Value,
+                    vec![PointerWrite {
+                        offset: 0,
+                        data: value_write_data.clone(),
+                    }],
+                ),
+                (
+                    PtrType::HashWithValue,
+                    vec![PointerWrite {
+                        offset: HASH_LEN,
+                        data: value_write_data.clone(),
+                    }],
+                ),
+                (
+                    PtrType::TwoHashesWithValue,
+                    vec![PointerWrite {
+                        offset: 2 * HASH_LEN,
+                        data: value_write_data,
+                    }],
+                ),
+            ]
+            .into_iter()
+            .collect(),
+        )
+    }
+}
+
+impl PointerWritesConfig {
+    pub fn empty() -> PointerWritesConfig {
+        PointerWritesConfig(HashMap::new())
+    }
+
+    /// Set the `PointerWrite`s for the specified pointer type.
+    pub fn set_rule(&mut self, ptr_type: PtrType, pointer_writes: Vec<PointerWrite>) {
```

Comment:
You see, the config is a public type that will be used by crate users to configurate wasm generation.

I believe letting them to work with `PointerWrite` is wrong. That's too complicated. If it's a map `PointerType -> Range`, then it will be decent enough. All other convertions should be done on the upper level, inside invocator or submodules.

Reply:
I assume that there will be some more variants in the `PointerWriteDataGenerator` that will use not only the ranges and/or ranges of another type, like `RangeInclusive<u64>` for `Gas`

Reply:
This is a really really bad ux for crate users to force them to instantiate `PointerWriteGenerator`. That offsets for different types of pointer can be pre-calculated on other levels, not on user level.

Reply:
#3398